File: Ex-2/java/NFAToDFA.java
import java.util.ArrayList;
import java.util.HashSet;

public class NFAToDFA{
    final static String epsilon = "epsilon";
    public static ArrayList<State> allStates = new ArrayList<State>();
    public static HashSet<Integer> allNames = new HashSet<Integer>();
    public static State epsilonClosure(NFA nfa, State state){
        HashSet<Edge> edgesFromNode;
        if(state.length() == 0){
            return state;
        }
        int name = Integer.parseInt(state.name.substring(1))+1;
        System.out.println("Debug - epsilonClosure\n"+name);
        while(allNames.contains(name)){
            name += 1;
        }
        allNames.add(name);
        State newState = new State("s"+name);
        for (int node: state.nodes){
            edgesFromNode = nfa.findEdgesFrom(node);
            for(Edge e : edgesFromNode){
                if(e.label.equalsIgnoreCase(epsilon)){
                    newState.addNode(e.to);
                }
            }
        }
        newState.extend(epsilonClosure(nfa, newState));
        newState.extend(state);
        return newState;
    }

    public static State closure(NFA nfa, State state, String label){
        HashSet<Edge> edgesFromNode;
        int name = Integer.parseInt(state.name.substring(1))+1;
        while(allNames.contains(name)){
            name+=1;
        }
        allNames.add(name);
        State newState = new State("s"+name);
        for (int i : state.nodes){
            edgesFromNode = nfa.findEdgesFrom(i);
            for( Edge j : edgesFromNode){
                if(j.label.equalsIgnoreCase(label)){
                    newState.addNode(j.to);
                }
            }
        }
        return newState;
    }

    public static DFA nfaToDFA(NFA nfa){
        System.out.println("Debug - nfaToDFA\n"+nfa);
        DFA dfa = new DFA();
        ArrayList<State> processed = new ArrayList<State>();
        State s0 = new State("s0");
        allStates.removeAll(allStates);
        allNames.removeAll(allNames);
        allStates.add(s0);
        allNames.add(0);
        s0.addNode(0);
        int k = 0;
        while(true){
            for(State i : allStates){
                System.out.println("Debug - nfaToDFA\n"+i);
                if(processed.contains(i)){
                    continue;
                }
                State temp1 = epsilonClosure(nfa, i);
                System.out.println("Debug - nfaToDFA\n"+temp1.nodes);
                System.out.println("Debug - nfaToDFA\n"+nfa.alphabets);
                for(String j : nfa.alphabets){
                    System.out.println("Debug - nfaToDFA\n"+j);
                    State temp2 = closure(nfa, temp1, j);
                    temp2 = epsilonClosure(nfa, temp2);
                    if(!allStates.contains(temp2)){
                        allStates.add(temp2);
                        dfa.addEdge(i, temp2, j);
                        System.out.println("Debug - nfaToDFA\n"+dfa);
                    }
                    else{
                        int ind = allStates.indexOf(temp2);
                        dfa.addEdge(i, allStates.get(ind), j);
                        System.out.println("Debug - nfaToDFA\n"+dfa);
                    }
                }
                processed.add(i);
            }
            if(allStates.size() == processed.size()){
                break;
            }
        }
        return dfa;
    }
}

File: Ex-2/python/nfa_to_dfa.py
from nfa_utilities import NFA, DFA

epsilon = 'epsilon'
all_states = []
all_names = []
class State:
	def __init__(self, name):
		self.nodes = []
		self.name = name
	
	def add_node(self, node):
		self.nodes.extend([node])
		self.nodes = list(set(self.nodes))

	def add_nodes(self, nodes):
		self.nodes.extend(nodes)
		self.nodes = list(set(self.nodes))


	def extend(self, state):
		self.nodes.extend(state.nodes)
		self.nodes = list(set(self.nodes))

	def __str__(self):
		#return 'Nodes: {}'.format(self.nodes)
		return self.name

	def __eq__(self, state):
		for n in state.nodes:
			if n not in self.nodes:
				return False
		if(len(state.nodes)==len(self.nodes)):
			return True
		else:
			return False

	def __len__(self):
		return len(self.nodes)

	def __lt__(self, other):
		return int(self.name[1:]) < int(other.name[1:])

def epsilon_closure(nfa, state):
	if(len(state)==0):
		return state
	global epsilon
	name = int(state.name[1:])+1
	while(name in all_names):
		name += 1
	all_names.extend([name])
	new_state = State('s{}'.format(name))
	for node in state.nodes:
		edges = nfa.find_edges_from(node)
		for e in edges:
			if(e.label == epsilon):
				new_state.add_node(e.t)
	new_state.extend(epsilon_closure(nfa, new_state))
	new_state.extend(state)
	return new_state

def closure(nfa, state, inp):
	i = 1
	name = int(state.name[1:])+i
	while(name in all_names):
		i+=1
		name = int(state.name[1:])+i
	all_names.extend([name])
	new_state = State('s{}'.format(name))
	for i in state.nodes:
		edges = nfa.find_edges_from(i)
		for j in edges:
			if j.label == inp:
				new_state.add_node(j.t)
	return new_state

		
def nfa_to_dfa(nfa):
	global all_states
	dfa = DFA()
	processed = []
	s0 = State('s0')
	s0.add_node(0)
	all_states = []
	all_states.extend([s0])
	all_names.extend(['s0'])
	k = 0
	while(True):
		for i in all_states:
			if i in processed:
				continue
			temp1 = epsilon_closure(nfa, i)
			for j in nfa.get_alphabet():
				temp2 = closure(nfa, temp1, j)
				temp2 = epsilon_closure(nfa, temp2)
				if temp2 not in all_states:
					all_states.extend([temp2])
					dfa.add_edge(i, temp2, j)
					#dfa = dfa + i.name + ' ' + j + ' ' + temp2.name + '\n'
				else:
					ind = all_states.index(temp2)
					dfa.add_edge(i, all_states[ind], j)
					#dfa = dfa + i.name + ' ' + j + ' ' + all_states[ind].name + '\n'
			processed.extend([i])
		
			
			"""print([[k.name, str(k)] for k in all_states])
			print([[k.name, str(k)] for k in processed])"""
		if(len(all_states) == len(processed)):
			break
	return dfa	
			